new/prusti-interface/src/specs/typed.rs

```diff
@@ -1,74 +1,124 @@
 /// A set of untyped specifications associated with a single element.
-pub type SpecificationSet = common::SpecificationSet<ExpressionId, BodyId, Arg>;
+pub type SpecificationSet<'tcx> = common::SpecificationSet<ExpressionId, LocalDefId, TyKind<'tcx>>;
 /// A set of untyped specifications associated with a loop.
-pub type LoopSpecification = common::LoopSpecification<ExpressionId, BodyId, Arg>;
+pub type LoopSpecification<'tcx> = common::LoopSpecification<ExpressionId, LocalDefId, TyKind<'tcx>>;
 /// A set of untyped specifications associated with a procedure.
-pub type ProcedureSpecification = common::ProcedureSpecification<ExpressionId, BodyId, Arg>;
+pub type ProcedureSpecification<'tcx> = common::ProcedureSpecification<ExpressionId, LocalDefId, TyKind<'tcx>>;
 /// A map of untyped specifications for a specific crate.
-pub type SpecificationMap = HashMap<common::SpecificationId, SpecificationSet>;
+pub type SpecificationMap<'tcx> = HashMap<common::SpecificationId, SpecificationSet<'tcx>>;
 /// An assertion that has no types associated with it.
-pub type Assertion = common::Assertion<ExpressionId, BodyId, Arg>;
+pub type Assertion<'tcx> = common::Assertion<ExpressionId, LocalDefId, TyKind<'tcx>>;
 /// An assertion kind that has no types associated with it.
-pub type AssertionKind = common::AssertionKind<ExpressionId, BodyId, Arg>;
+pub type AssertionKind<'tcx> = common::AssertionKind<ExpressionId, LocalDefId, TyKind<'tcx>>;
 /// An expression that has no types associated with it.
-pub type Expression = common::Expression<ExpressionId, BodyId>;
-/// A trigger set that has not types associated with it.
-pub type TriggerSet = common::TriggerSet<ExpressionId, BodyId>;
+pub type Expression = common::Expression<ExpressionId, LocalDefId>;
+/// A trigger set that has no types associated with it.
+pub type TriggerSet = common::TriggerSet<ExpressionId, LocalDefId>;
+/// For all variables that have no types associated with it.
+pub type ForAllVars<'tcx> = common::ForAllVars<ExpressionId, TyKind<'tcx>>;
+/// A trigger that has no types associated with it.
+pub type Trigger = common::Trigger<ExpressionId, LocalDefId>;
 
-pub trait StructuralToTyped<Target> {
-    fn to_typed(self, typed_expressions: &HashMap<String, rustc_hir::BodyId>) -> Target;
+pub trait StructuralToTyped<'tcx, Target> {
+    fn to_typed(self, typed_expressions: &HashMap<String, HirId>, tcx: TyCtxt<'tcx>) -> Target;
 }
 
-impl StructuralToTyped<Expression> for json::Expression {
-    fn to_typed(self, typed_expressions: &HashMap<String, rustc_hir::BodyId>) -> Expression {
-        let body_id = typed_expressions[&format!("{}_{}", self.spec_id, self.expr_id)];
+impl<'tcx> StructuralToTyped<'tcx, Expression> for json::Expression {
+    fn to_typed(self, typed_expressions: &HashMap<String, HirId>, tcx: TyCtxt<'tcx>) -> Expression {
+        let hir_id = typed_expressions[&format!("{}_{}", self.spec_id, self.expr_id)];
```

Comment:
Wouldn't it be better to store `LocalDefId` in `typed_expressions`, which would allow avoiding the `tcx` argument here? (Visitor is created from `tcx`, so it seems more natural to me to have the argument there.)

Reply:
but then you need the `tcx` to get the `Body`: `body = tcx.mir_validated(local_def_id)`

Reply:
Yes, but that is the third piece of code which needs `tcx` anyway.

Reply:
how would this work then?
```rust
fn to_typed(self, typed_expressions: &HashMap<String, HirId>, tcx: TyCtxt<'tcx>) -> ForAllVars<'tcx> {
        let hir_id = typed_expressions[&format!("{}_{}", self.spec_id, self.expr_id)];
        let local_id = tcx.hir().local_def_id(hir_id);
        let (body, _) = tcx.mir_validated(local_id);
```

Reply:
True. However, I think that it would still be better to store `LocalDefId` instead of `HirId`.

Reply:
That's something we can probably do. I will change it.

Reply:
Thanks!